sktime/dists_kernels/scipy_dist.py

```diff
@@ -0,0 +1,80 @@
+
+from scipy.spatial.distance import cdist
+
+from sktime.dists_kernels._base import BasePairwiseTransformer
+
+
+class ScipyDist(BasePairwiseTransformer):
+    """
+    computes pairwise distances using scipy.spatial.distance.cdist
+        includes Euclidean distance and p-norm (Minkowski) distance
+            note: weighted distances are not supported
+
+    Hyper-parameters:
+        metric: string or function, as in cdist; default = 'euclidean'
+            if string, one of: 'braycurtis', 'canberra', 'chebyshev', 'cityblock',
+                'correlation', 'cosine', 'dice', 'euclidean', 'hamming', 'jaccard',
+                'jensenshannon', 'kulsinski', 'mahalanobis', 'matching', 'minkowski',
+                'rogerstanimoto', 'russellrao', 'seuclidean', 'sokalmichener',
+                'sokalsneath', 'sqeuclidean', 'yule'
+            if function, should have signature 1D-np.array x 1D-np.array -> float
+        p: if metric='minkowski', the "p" in "p-norm", otherwise irrelevant
+        colalign: string, one of 'intersect' (default), 'force-align', 'none'
+            controls column alignment if X, X2 passed in fit are pd.DataFrame
+            columns between X and X2 are aligned via column names
+            if 'intersect', distance is computed on columns occurring both in X and X2,
+                other columns are discarded; column ordering in X2 is copied from X
+            if 'force-align', raises an error if the set of columns in X, X2 differs
+            if 'none', X and X2 are passed through unmodified (no columns are aligned)
+                note: this will potentially align "non-matching" columns
+    """
+
+    def __init__(self, metric="euclidean", p=2, colalign="intersect"):
+
+        self.metric = metric
+        self.p = p
+        self.colalign = colalign
+
+        super(ScipyDist, self).__init__()
+
+    def _transform(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2
+                if X2 is not passed, is equal to X
+                if X/X2 is a pd.DataFrame and contains non-numeric columns,
+                    these are removed before computation
+
+        Args:
+            X: pd.DataFrame of length n, or 2D np.array of 'float' with n rows
+
+        Optional args:
+            X2: pd.DataFrame of length m, or 2D np.array of 'float' with m rows
+
+        Returns:
+            distmat: np.array of shape [n, m]
+                (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+                    (non-numeric columns are removed before for DataFrame X/X2)
+        """
+
+        p = self.p
+        metric = self.metric
+
+        if isinstance(X, pd.DataFrame):
+            X = X.select_dtypes("number").to_numpy(dtype="float")
+
+        if isinstance(X2, pd.DataFrame):
+            X2 = X2.select_dtypes("number").to_numpy(dtype="float")
+
+        distmat = cdist(XA=X, XB=X2, metric=metric, p=p)
```

Comment:
So potententially I have something wrong locally but when I try run this code passing the parameters p=p to cdist it causes a TypeError. Which is interesting because it defaults to 2 (which is the valueu being passed as p set by your default parameter in the constructor). So I don't know why this is caused. Removing the p parameter it runs fine (even thought cdist defaults p to 2).

Reply:
strange, let me try this locally again. What is the message in your `TypeError`?

Reply:
So something that I found that doesn't seem intentional from cdist is that passing p throws an error using any other metric than "minkowski". Even passing p=None throws an error. I would assume this is a bug because surely you just ignore the parameter when you're not using minkowski

Reply:
yes, there was a bug which I fixed if you didn't pass `X2`.
It was in the base class `transform` input logic:

```python
    if X2 is None:
        X = X2
```
(old state)

d'oh

Reply:
O good spot I missed that too 😂

Reply:
is this your problem? If not, please paste code

Reply:
Sorry @fkiraly just got around to testing again. Same issue the problem is coming from cdist itself when you pass p as a parameter when it isnt used (i.e. for any other metric not 'minkowski').

Just to be more explicit with the error so you can hopefully recreate and see what I mean. Here is (and sorry for no syntax highlighting this is done in just terminal to get it in complete isolation), this is what im on about

![image](https://user-images.githubusercontent.com/4674372/124207454-79249400-dadd-11eb-93d8-413e581e8f49.png)

As you can see the function was working fine with euclidean until I added "p" as a parameter where it threw an error as euclidean doesn't use p. Therefore if you're not using "minkowski" as the metric you can't pass p as a parameter or an error is thrown.

My current work around is this:

```
def _transform(self, X, X2=None):
    p = self.p
    metric = self.metric
    
    if isinstance(X, pd.DataFrame):
      X = X.select_dtypes("number").to_numpy(dtype="float")
    
    if isinstance(X2, pd.DataFrame):
      X2 = X2.select_dtypes("number").to_numpy(dtype="float")
    
    if metric is "minkowski":
      distmat = cdist(XA=X, XB=X2, metric=metric, p=p)
    else:
      distmat = cdist(XA=X, XB=X2, metric=metric)
    
    return distmat
```

Reply:
this is resolved via @chrisholder's PR